### ISSUE TITLE
add support for running entirely in single precision

### ISSUE THREAD

**Issue by BenWibking** [quokka-astro/quokka#666]
**Describe the proposal**
Support running `amrex::Real` as `float` (compile with `-DAMReX_PRECISION=SINGLE`) and replace all uses of `double` with `amrex::Real`. Ensure that all tests still pass.

**Describe alternatives you've considered**
We have so far relied on users running on GPUs with hardware support for double precision arithmetic.

**Additional context**
Due to AI, future-generation GPUs may completely drop hardware support for double precision arithmetic.

**Comment by chongchonghe** [comment]
I totally agree and support this plan. I recently learned about another code under development that uses mixed precision and achieves extremely high speed, about 100 Mupdates/s/GPU on game cards (RTX 3080/3090). They use single precision for the approximate Riemann solver but use double precision in the conservation step.

**Comment by BenWibking** [comment]
That's interesting.

Over the next couple of years, the code needs to run without double precision at all. If we need extra precision, we can do Kahan summation or represent numbers as the sum of two single precision values (then overload arithmetic operators such as `+` and `*` to emulate double precision).

**Comment by chongchonghe** [comment]
Now that Claude is so powerful, it can probably create a PR to do this.

**Comment by chongchonghe** [comment]
@claude Evaluate the feasibility of creating a PR to do the task proposed in this issue. Further note that, since all unannotated floating point constants are double in C++, we also have to add the suffix _rt to every floating point constant that we use in the code.

**Comment by chongchonghe** [comment]
@BenWibking @markkrumholz I think Claude can probably do this job with some human help, since it's all systematic. There is a design consideration. If we decide to use single precision in the future, do we want to use single precision for everything, or mixed precision between hydro, radiation, and particles. I know AMReX has a separate parameter for particle precision, but I don't know if it allows setting two precision types in hydro and radiation, although we can always define our own (like `quokka::Real_rad`.

**Comment by BenWibking** [comment]
That's a good idea. We'd have to create our own type for radiation precision.
